File: phases/07-transformers-deep-dive/02-self-attention-from-scratch/code/main.jl
"""
从零实现自注意力机制 (Self-Attention from Scratch) —— Julia 版本

本模块实现了 Transformer 的核心组件——自注意力机制，包括：
- 数值稳定的 softmax
- 缩放点积注意力 (Scaled Dot-Product Attention)
- 单头自注意力 (SelfAttention)
- 多头自注意力 (MultiHeadSelfAttention)

核心公式：Attention(Q, K, V) = softmax(Q @ K^T / sqrt(d_k)) @ V

参考文献：
- Vaswani et al., "Attention Is All You Need" (2017)
  https://arxiv.org/abs/1706.03762
"""

using Random
using LinearAlgebra
using Printf


"""
    softmax_rows(M) -> Matrix

数值稳定的逐行 softmax。

对每行先减去最大值（防止 exp 溢出），再计算 exp 和归一化：
    softmax(x_i) = exp(x_i - max(x)) / Σ exp(x_j - max(x))
"""
function softmax_rows(M::Matrix{Float64})::Matrix{Float64}
    out = similar(M)
    for i in 1:size(M, 1)
        row = M[i, :]
        m = maximum(row)       # 找到行最大值，用于数值稳定
        e = exp.(row .- m)     # 减去最大值后取 exp
        s = sum(e)
        out[i, :] = e ./ s    # 归一化为概率分布
    end
    return out
end


"""
    scaled_dot_product_attention(Q, K, V) -> (output, weights)

缩放点积注意力 (Scaled Dot-Product Attention)。

核心公式：Attention(Q, K, V) = softmax(Q @ K^T / sqrt(d_k)) @ V

缩放因子 1/sqrt(d_k) 的作用：防止点积值过大导致 softmax 梯度消失。

参数:
- Q: Query 矩阵，形状 (seq_len, dk)
- K: Key 矩阵，形状 (seq_len, dk)
- V: Value 矩阵，形状 (seq_len, dv)

返回:
- output: 注意力加权后的输出，形状 (seq_len, dv)
- weights: 注意力权重矩阵，形状 (seq_len, seq_len)
"""
function scaled_dot_product_attention(Q::Matrix{Float64}, K::Matrix{Float64},
                                      V::Matrix{Float64})
    dk = size(Q, 2)
    scores = (Q * transpose(K)) ./ sqrt(dk)  # Q @ K^T / sqrt(d_k)
    weights = softmax_rows(scores)             # softmax 归一化
    output = weights * V                       # 注意力加权求和
    return output, weights
end


"""
单头自注意力层 (Single-Head Self-Attention)。

将输入 X 分别通过三个线性变换得到 Q、K、V，然后计算缩放点积注意力。

字段:
- Wq: Query 权重矩阵，形状 (d_model, dk)
- Wk: Key 权重矩阵，形状 (d_model, dk)
- Wv: Value 权重矩阵，形状 (d_model, dv)
- dk: Query/Key 投影维度
"""
struct SelfAttention
    Wq::Matrix{Float64}
    Wk::Matrix{Float64}
    Wv::Matrix{Float64}
    dk::Int
end


"""
    SelfAttention(d_model, dk, dv; seed=42)

构造函数，使用 Xavier 初始化权重矩阵。
Xavier 初始化：scale = sqrt(2 / (fan_in + fan_out))，保持前向传播方差稳定。
"""
function SelfAttention(d_model::Int, dk::Int, dv::Int; seed::Int=42)
    rng = MersenneTwister(seed)
    # Xavier 初始化：保持前向传播方差稳定
    scale_qk = sqrt(2.0 / (d_model + dk))
    scale_v = sqrt(2.0 / (d_model + dv))
    Wq = scale_qk .* randn(rng, d_model, dk)  # Query 权重
    Wk = scale_qk .* randn(rng, d_model, dk)  # Key 权重
    Wv = scale_v .* randn(rng, d_model, dv)    # Value 权重
    return SelfAttention(Wq, Wk, Wv, dk)
end


"""
    forward(attn, X) -> (output, weights)

前向传播：Q = X @ Wq, K = X @ Wk, V = X @ Wv，然后计算注意力。

参数:
- attn: SelfAttention 层
- X: 输入矩阵，形状 (seq_len, d_model)
"""
function forward(attn::SelfAttention, X::Matrix{Float64})
    Q = X * attn.Wq  # Query：表示"我在找什么"
    K = X * attn.Wk  # Key：表示"我有什么信息"
    V = X * attn.Wv  # Value：表示"我能提供什么"
    return scaled_dot_product_attention(Q, K, V)
end


"""
多头自注意力层 (Multi-Head Self-Attention)。

将注意力计算拆分到多个"头"并行进行，每个头可以学习不同的注意力模式。
流程：
1. 将输入投影到 n_heads 个子空间，每个头独立计算注意力
2. 拼接所有头的输出
3. 通过输出投影矩阵 Wo 融合多头信息

字段:
- heads: SelfAttention 层的向量
- Wo: 输出投影矩阵，形状 (n_heads * dv, d_model)
- n_heads: 注意力头的数量
"""
struct MultiHeadSelfAttention
    heads::Vector{SelfAttention}
    Wo::Matrix{Float64}
    n_heads::Int
end


"""
    MultiHeadSelfAttention(d_model, n_heads; seed=42)

构造函数，创建 n_heads 个独立的单头注意力层和输出投影矩阵。
"""
function MultiHeadSelfAttention(d_model::Int, n_heads::Int; seed::Int=42)
    @assert n_heads > 0 "n_heads must be > 0"
    @assert d_model > 0 "d_model must be > 0"
    @assert d_model % n_heads == 0 "d_model must be divisible by n_heads"
    dk = d_model ÷ n_heads  # 每个头的维度
    dv = d_model ÷ n_heads
    # 创建 n_heads 个独立的单头注意力层
    heads = [SelfAttention(d_model, dk, dv; seed=seed + i) for i in 1:n_heads]
    rng = MersenneTwister(seed + n_heads + 1)
    # 输出投影矩阵：将拼接后的多头输出映射回 d_model 维
    scale = sqrt(2.0 / (d_model + d_model))
    Wo = scale .* randn(rng, n_heads * dv, d_model)
    return MultiHeadSelfAttention(heads, Wo, n_heads)
end


"""
    forward(mha, X) -> (output, weights_per_head)

前向传播：计算多头自注意力。

1. 并行计算每个头的注意力
2. 拼接所有头的输出
3. 通过输出投影矩阵 Wo 融合
"""
function forward(mha::MultiHeadSelfAttention, X::Matrix{Float64})
    head_outputs = Matrix{Float64}[]
    weights_per_head = Matrix{Float64}[]
    # 并行计算每个头的注意力
    for head in mha.heads
        out, w = forward(head, X)
        push!(head_outputs, out)
        push!(weights_per_head, w)
    end
    # 拼接所有头的输出并投影
    concat = hcat(head_outputs...)
    return concat * mha.Wo, weights_per_head
end


"""
打印注意力权重矩阵。
"""
function print_attention_matrix(weights::Matrix{Float64}, tokens::Vector{String})
    # 列标题
    print("\n      ")
    for token in tokens
        @printf("%6s", token)
    end
    println()
    # 每行：行标签 + 权重值
    for i in 1:length(tokens)
        @printf("%6s", tokens[i])
        for j in 1:length(tokens)
            @printf("%6.3f", weights[i, j])
        end
        println()
    end
end


"""
用 ASCII 字符绘制注意力权重热力图。
权重值越大，使用的字符越"密"。
"""
function ascii_heatmap(weights::Matrix{Float64}, tokens::Vector{String};
                       chars::String=" .:-=+*#%@")
    # 列标题
    print("\n      ")
    for t in tokens
        @printf("%6s", t)
    end
    println()
    w_max = maximum(weights)
    # 热力图主体
    for i in 1:length(tokens)
        @printf("%6s", tokens[i])
        for j in 1:length(tokens)
            # 将权重值映射到字符索引
            level = Int(floor(weights[i, j] * (length(chars) - 1) / w_max))
            level = min(level, length(chars) - 1)
            ch = chars[level + 1]
            @printf("    %s ", ch)
        end
        println()
    end
end


"""
Softmax 数值稳定性演示。
展示减去最大值后，即使输入值很大也不会溢出。
"""
function demo_softmax_stability()
    println("\n" * "=" ^ 60)
    println("SOFTMAX NUMERIC STABILITY")
    println("=" ^ 60)
    # 普通 logits
    logits = reshape([2.0, 1.0, 0.1], 1, 3)
    probs = softmax_rows(logits)
    @printf("\nLogits:  [%s]\n", join([@sprintf("%.4f", v) for v in logits], ", "))
    @printf("Softmax: [%s]\n", join([@sprintf("%.4f", v) for v in probs], ", "))
    @printf("Sum:     %.4f\n", sum(probs))

    # 大数值 logits：展示数值稳定性
    big_logits = reshape([100.0, 200.0, 300.0], 1, 3)
    big_probs = softmax_rows(big_logits)
    @printf("\nLarge logits:  [%s]\n",
            join([@sprintf("%.1f", v) for v in big_logits], ", "))
    @printf("Softmax:       [%s]\n",
            join([@sprintf("%.4f", v) for v in big_probs], ", "))
    @printf("Sum:           %.4f\n", sum(big_probs))
    println("(no overflow because we subtract the row maximum before exp)")
end


"""
演示单头自注意力：用随机向量模拟词嵌入，计算注意力权重并可视化。
"""
function demo_self_attention()
    println("=" ^ 60)
    println("SELF-ATTENTION FROM SCRATCH")
    println("=" ^ 60)

    tokens = ["The", "cat", "sat", "on", "the", "mat"]
    n_tokens = length(tokens)
    d_model = 16  # 嵌入维度
    dk = 8        # Query/Key 投影维度
    dv = 8        # Value 投影维度

    # 用随机向量模拟词嵌入（实际中由 Embedding 层生成）
    rng = MersenneTwister(42)
    X = randn(rng, n_tokens, d_model)

    @printf("\nSentence: %s\n", join(tokens, " "))
    @printf("Tokens: %d  d_model: %d  dk: %d  dv: %d\n", n_tokens, d_model, dk, dv)
    @printf("Input shape: (%d, %d)\n", size(X, 1), size(X, 2))

    # 创建自注意力层并计算前向传播
    attn = SelfAttention(d_model, dk, dv; seed=42)
    output, weights = forward(attn, X)
    @printf("\nOutput shape: (%d, %d)\n", size(output, 1), size(output, 2))
    println("\nAttention weights:")
    print_attention_matrix(weights, tokens)
    println("\nASCII heatmap (denser char = higher attention):")
    ascii_heatmap(weights, tokens)
    return tokens, X, d_model
end


"""
演示多头自注意力：展示不同头学到不同的注意力模式。
"""
function demo_multi_head(tokens::Vector{String}, X::Matrix{Float64}, d_model::Int)
    println("\n" * "=" ^ 60)
    println("MULTI-HEAD SELF-ATTENTION")
    println("=" ^ 60)
    n_heads = 2
    mha = MultiHeadSelfAttention(d_model, n_heads; seed=42)
    out, head_weights = forward(mha, X)
    @printf("\nHeads: %d  Output shape: (%d, %d)\n",
            n_heads, size(out, 1), size(out, 2))
    # 打印每个头的注意力权重，展示不同头学到不同模式
    for (h, w) in enumerate(head_weights)
        @printf("\nHead %d attention weights:\n", h)
        print_attention_matrix(w, tokens)
    end
end


function main()
    tokens, X, d_model = demo_self_attention()
    demo_multi_head(tokens, X, d_model)
    demo_softmax_stability()
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: phases/07-transformers-deep-dive/02-self-attention-from-scratch/code/main.rs
//! 从零实现自注意力机制 (Self-Attention from Scratch) —— Rust 版本
//!
//! 仅使用标准库，手动实现行优先矩阵运算和缩放点积注意力。
//! 适合理解注意力机制的底层计算细节。
//!
//! 核心公式：Attention(Q, K, V) = softmax(Q @ K^T / sqrt(d_k)) @ V
//!
//! 参考文献：
//!   - Vaswani 2017, "Attention Is All You Need": https://arxiv.org/abs/1706.03762
//!   - candle reference attention kernel: https://github.com/huggingface/candle/blob/main/candle-nn/src/ops.rs
//!   - Karpathy llm.c attention forward pass: https://github.com/karpathy/llm.c/blob/master/train_gpt2.c
//!
//! 编译运行: rustc --edition 2021 main.rs -o /tmp/sa && /tmp/sa

use std::f32::consts::E;

/// 行优先 (row-major) 矩阵，底层存储为一维 Vec<f32>。
///
/// 索引方式：data[i * cols + j] 访问第 i 行第 j 列元素。
/// 这是大多数深度学习框架（如 PyTorch 默认）采用的内存布局。
struct Mat {
    rows: usize,
    cols: usize,
    data: Vec<f32>,
}

impl Mat {
    /// 创建全零矩阵
    fn zeros(rows: usize, cols: usize) -> Self {
        Mat { rows, cols, data: vec![0.0; rows * cols] }
    }

    /// 按 (行, 列) 读取元素
    #[inline] fn at(&self, i: usize, j: usize) -> f32 { self.data[i * self.cols + j] }
    /// 按 (行, 列) 写入元素
    #[inline] fn set(&mut self, i: usize, j: usize, v: f32) { self.data[i * self.cols + j] = v; }

    /// 矩阵乘法：self @ b
    ///
    /// 使用 i-k-j 循环顺序，对稀疏 A 矩阵跳过零元素以优化性能。
    /// 要求 self.cols == b.rows。
    fn matmul(&self, b: &Mat) -> Mat {
        assert_eq!(self.cols, b.rows, "shape mismatch: {}x{} @ {}x{}", self.rows, self.cols, b.rows, b.cols);
        let mut out = Mat::zeros(self.rows, b.cols);
        for i in 0..self.rows {
            for k in 0..self.cols {
                let aik = self.at(i, k);
                if aik == 0.0 { continue; }  // 跳过零元素，优化稀疏矩阵
                let row_base = i * out.cols;
                let bk_base = k * b.cols;
                for j in 0..b.cols {
                    out.data[row_base + j] += aik * b.data[bk_base + j];
                }
            }
        }
        out
    }

    /// 矩阵转置：返回 self^T
    fn transpose(&self) -> Mat {
        let mut t = Mat::zeros(self.cols, self.rows);
        for i in 0..self.rows {
            for j in 0..self.cols {
                t.set(j, i, self.at(i, j));
            }
        }
        t
    }

    /// 标量乘法：原地将所有元素乘以 s
    fn scale(&mut self, s: f32) {
        for v in self.data.iter_mut() { *v *= s; }
    }
}

/// 数值稳定的逐行 softmax。
///
/// 算法：对每行先减去最大值（防止 exp 溢出），再计算 exp 和归一化。
/// softmax(x_i) = exp(x_i - max(x)) / Σ exp(x_j - max(x))
fn softmax_rows(m: &Mat) -> Mat {
    let mut out = Mat::zeros(m.rows, m.cols);
    for i in 0..m.rows {
        // 找到当前行的最大值
        let mut row_max = f32::NEG_INFINITY;
        for j in 0..m.cols { if m.at(i, j) > row_max { row_max = m.at(i, j); } }
        // 计算 exp(x - max) 并累加求和
        let mut sum = 0.0f32;
        for j in 0..m.cols {
            let e = E.powf(m.at(i, j) - row_max);
            out.set(i, j, e);
            sum += e;
        }
        // 归一化：每个元素除以总和
        let inv = 1.0 / sum;
        for j in 0..m.cols {
            let v = out.at(i, j) * inv;
            out.set(i, j, v);
        }
    }
    out
}

/// 缩放点积注意力 (Scaled Dot-Product Attention)。
///
/// 核心公式：Attention(Q, K, V) = softmax(Q @ K^T / sqrt(d_k)) @ V
///
/// 步骤：
/// 1. 计算注意力分数 scores = Q @ K^T
/// 2. 缩放 scores = scores / sqrt(d_k)（防止点积过大导致 softmax 梯度消失）
/// 3. 归一化 weights = softmax(scores)
/// 4. 加权求和 output = weights @ V
fn scaled_dot_product_attention(q: &Mat, k: &Mat, v: &Mat) -> (Mat, Mat) {
    let dk = q.cols as f32;
    let k_t = k.transpose();
    let mut scores = q.matmul(&k_t);       // Q @ K^T
    scores.scale(1.0 / dk.sqrt());          // 除以 sqrt(d_k) 缩放
    let weights = softmax_rows(&scores);    // softmax 归一化
    let out = weights.matmul(v);            // 注意力加权求和
    (out, weights)
}

/// 确定性伪随机数生成器（无外部依赖）。
///
/// 使用 Lehmer LCG (线性同余生成器) + Box-Muller 变换生成高斯随机数。
/// - LCG: state = state * a + c，取高 31 位作为均匀随机数
/// - Box-Muller: 将两个均匀随机数转换为标准正态分布
struct Rng { state: u64 }
impl Rng {
    fn new(seed: u64) -> Self { Rng { state: seed.wrapping_mul(0x9E37_79B9_7F4A_7C15) | 1 } }
    /// 生成下一个 u32 随机数（取 state 的高 31 位）
    fn next_u32(&mut self) -> u32 {
        self.state = self.state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        (self.state >> 33) as u32
    }
    /// 生成 (0, 1) 区间的均匀分布随机数
    fn uniform(&mut self) -> f32 {
        (self.next_u32() as f32 + 1.0) / (u32::MAX as f32 + 2.0)
    }
    /// Box-Muller 变换：将两个均匀随机数转换为标准正态分布 N(0,1)
    fn gauss(&mut self) -> f32 {
        let u1 = self.uniform();
        let u2 = self.uniform();
        (-2.0 * u1.ln()).sqrt() * (2.0 * std::f32::consts::PI * u2).cos()
    }
}

/// 生成指定形状的高斯随机矩阵，元素 ~ N(0, scale²)
fn randn(rows: usize, cols: usize, scale: f32, rng: &mut Rng) -> Mat {
    let mut m = Mat::zeros(rows, cols);
    for v in m.data.iter_mut() { *v = rng.gauss() * scale; }
    m
}

/// 单头自注意力层 (Single-Head Self-Attention)。
///
/// 将输入 X 通过三个线性变换得到 Q、K、V，然后计算缩放点积注意力。
/// 权重矩阵使用 Xavier 初始化：scale = sqrt(2 / (fan_in + fan_out))
struct SelfAttention {
    wq: Mat,  // Query 权重矩阵：(d_model, dk)
    wk: Mat,  // Key 权重矩阵：(d_model, dk)
    wv: Mat,  // Value 权重矩阵：(d_model, dv)
}

impl SelfAttention {
    fn new(d_model: usize, dk: usize, dv: usize, rng: &mut Rng) -> Self {
        // Xavier 初始化：保持前向传播方差稳定
        let s_qk = (2.0 / (d_model + dk) as f32).sqrt();
        let s_v = (2.0 / (d_model + dv) as f32).sqrt();
        SelfAttention {
            wq: randn(d_model, dk, s_qk, rng),
            wk: randn(d_model, dk, s_qk, rng),
            wv: randn(d_model, dv, s_v, rng),
        }
    }

    /// 前向传播：Q = X @ Wq, K = X @ Wk, V = X @ Wv，然后计算注意力
    fn forward(&self, x: &Mat) -> (Mat, Mat) {
        let q = x.matmul(&self.wq);  // Query：表示"我在找什么"
        let k = x.matmul(&self.wk);  // Key：表示"我有什么信息"
        let v = x.matmul(&self.wv);  // Value：表示"我能提供什么"
        scaled_dot_product_attention(&q, &k, &v)
    }
}

/// 打印注意力权重矩阵
fn print_attention(weights: &Mat, tokens: &[&str]) {
    // 列标题
    print!("      ");
    for t in tokens { print!("{:>7}", t); }
    println!();
    // 每行：行标签 + 权重值
    for i in 0..weights.rows {
        print!("{:>6}", tokens[i]);
        for j in 0..weights.cols { print!("{:>7.3}", weights.at(i, j)); }
        println!();
    }
}

/// 用 ASCII 字符绘制注意力权重热力图
///
/// 字符从稀到密：' ' < ░ < ▒ < ▓ < █，权重越大字符越密
fn ascii_heatmap(weights: &Mat, tokens: &[&str]) {
    let chars = [' ', '\u{2591}', '\u{2592}', '\u{2593}', '\u{2588}'];
    // 找到最大权重值用于归一化
    let mut w_max = 0.0f32;
    for v in &weights.data { if *v > w_max { w_max = *v; } }
    // 列标题
    print!("      ");
    for t in tokens { print!("{:>7}", t); }
    println!();
    // 热力图主体
    for i in 0..weights.rows {
        print!("{:>6}", tokens[i]);
        for j in 0..weights.cols {
            // 将权重值映射到字符索引
            let level = ((weights.at(i, j) * (chars.len() - 1) as f32) / w_max) as usize;
            let level = level.min(chars.len() - 1);
            print!("     {} ", chars[level]);
        }
        println!();
    }
}

/// 数值稳定的 softmax（一维向量版本，用于演示）
fn softmax_vec(logits: &[f32]) -> Vec<f32> {
    let mut m = f32::NEG_INFINITY;
    for &x in logits { if x > m { m = x; } }
    let exps: Vec<f32> = logits.iter().map(|x| (x - m).exp()).collect();
    let s: f32 = exps.iter().sum();
    exps.into_iter().map(|x| x / s).collect()
}

fn main() {
    // === 示例句子 ===
    let sentence = ["The", "cat", "sat", "on", "the", "mat"];
    let n_tokens = sentence.len();
    let d_model: usize = 16;  // 嵌入维度
    let dk: usize = 8;        // Query/Key 投影维度
    let dv: usize = 8;        // Value 投影维度

    println!("{}", "=".repeat(60));
    println!("SELF-ATTENTION FROM SCRATCH (Rust port)");
    println!("{}", "=".repeat(60));

    // 用随机向量模拟词嵌入（实际中由 Embedding 层生成）
    let mut rng = Rng::new(42);
    let x = randn(n_tokens, d_model, 1.0, &mut rng);
    println!("\nSentence: {}", sentence.join(" "));
    println!("Tokens: {}, d_model: {}, dk: {}, dv: {}", n_tokens, d_model, dk, dv);
    println!("Input shape: ({}, {})", x.rows, x.cols);

    // 创建自注意力层并计算前向传播
    let mut rng_w = Rng::new(42);
    let attn = SelfAttention::new(d_model, dk, dv, &mut rng_w);
    let (out, weights) = attn.forward(&x);

    println!("\nOutput shape: ({}, {})", out.rows, out.cols);
    println!("\nAttention weights:");
    print_attention(&weights, &sentence);

    println!("\nASCII heatmap (darker = higher attention):");
    ascii_heatmap(&weights, &sentence);

    // === Softmax 数值稳定性演示 ===
    println!("\n{}", "=".repeat(60));
    println!("SOFTMAX DEMO");
    println!("{}", "=".repeat(60));

    // 普通 logits
    let logits = [2.0f32, 1.0, 0.1];
    let probs = softmax_vec(&logits);
    println!("\nLogits:  {:?}", logits);
    println!("Softmax: {:?}", probs.iter().map(|p| (p * 10000.0).round() / 10000.0).collect::<Vec<_>>());
    println!("Sum:     {:.4}", probs.iter().sum::<f32>());

    // 大数值 logits：展示数值稳定性（减去 max 后不会溢出）
    let large = [100.0f32, 200.0, 300.0];
    let probs_l = softmax_vec(&large);
    println!("\nLarge logits:  {:?}", large);
    println!("Softmax:       {:?}", probs_l.iter().map(|p| (p * 10000.0).round() / 10000.0).collect::<Vec<_>>());
    println!("Sum:           {:.4}", probs_l.iter().sum::<f32>());
    println!("(numerically stable, no overflow)");

    // === 性能基准测试 ===
    println!("\n{}", "=".repeat(60));
    println!("MICROBENCH: 10K attention forwards");
    println!("{}", "=".repeat(60));
    let start = std::time::Instant::now();
    let mut sink = 0.0f32;
    for _ in 0..10_000 {
        let (o, _) = attn.forward(&x);
        sink += o.at(0, 0);  // 防止编译器优化掉整个循环
    }
    let elapsed = start.elapsed();
    println!("10K forwards in {:.2}ms ({:.0}/sec)  sink={:.4}",
        elapsed.as_secs_f64() * 1000.0,
        10_000.0 / elapsed.as_secs_f64(),
        sink,
    );
}
